### ISSUE TITLE
Give plugin writers a hook for requesting user information

### ISSUE THREAD

**Issue by aaronchantrill** [NaomiProject/Naomi#182]
## Detailed Description
Before we can really expect people to start building plugins for Naomi, we need to provide a consistent interface for plugin developers. One of the areas where we especially need effort is creating an API for gathering information from the user that is required for the plugin.

Currently, if a developer needs to gather some information from the user, they have two choices:
1. Tell the user to open their profile.yml file and manually add the required information
2. Add a request for the information to the naomi/populate.py program and ask the user to run ./Naomi.py with the --repopulate flag.

Both of these choices are overly complicated for the user. The first requires the user to locate and edit a hidden file, provides no assistance to the user, and provides no error or sanity checking. The second basically requires the plugin to become an official part of Naomi, requires the user to know how to run Naomi with flags, and forces the user to go through all of the options for setting up Naomi.

I propose an api for gathering missing information from the user. The author of the plugin will define a data structure defining the following specifications:

**NAME** - (required) The name of the variable as seen by the developer. Since this can be a multi-level variable, it will be passed as a tuple containing the nodes to the path. So, to set the value

```
email:
  address: me@myhostname.net
```

you would use the tuple ('email','address'). This will form the key for a dictionary containing the whole form.

**TYPE** - (optional, default "textbox") - This is the general form for the control, without specifics about how to implement it. These fall into the following general categories:
* A simple text input box (default)
* A simple yes/no question (only possible responses are yes or no)
* A combobox where the user can select a value from a list or enter a new value
* A listbox where the user can only select a value from the provided list (or an empty string)
* A file select box where the user can either browse the server filesystem looking for a file, or can upload a file from their local filesystem which will be saved in a location on the server.

Notice that I don't specify radio buttons or checkboxes since these are implementation details covered by the concept of a listbox. We will probably also need a multiple list box where multiple options can be selected, but there isn't an instance where that is required in the current populate.py program.

**TITLE** - (required) A brief (five words or so) description of the setting.

**DESCRIPTION** - (optional, default "") a longer description of what exactly the setting does.

**DEFAULT** - (optional, default "") The default value as chosen by the developer of the plugin. This can be a function. This will only show up if the user has not previously selected a value.

**OPTIONS** - (optional, default []) In the case of a combo or list box, a list containing the available choices. This can be a function.

**VALIDATION** - (optional, default True) a function used to validate the user response and make sure the user entry is valid. If the value entered by the user is not valid, then the setting will not be saved. On the command line, the question will continue to be put to the user until they enter a valid response. An empty response will always be valid and can be taken to mean that the user has chosen not to respond. Some standard validation functions (for checking things like email address, number, integer, etc.) should be provided for developers.

**ACTIVE** - (optional, default True) a function used to tell if this should be active or not based on the current state of the form. This is used to disable or enable an option depending on values entered into the system.

Example: Email or Text Message

This is a sequence from the current populate.py. I believe that it is for the "notifications" plugin. Here is what the definition would look like:

```
settings = OrderedDict(
    [
        (
            ('email','address'),{
                'type':'textbox',
                'title':'Please enter your email address',
                'description':"".format([
                    'This email address can be used by the check_email plugin',
                    'to read your emails to you. It can also be used by the',
                    'notifications plugin to send notifications to you.'
                ]),
                'validation':lambda Email : True if re.match('^[^@]+@[^@]+\.[^@\.]+$') else False
            }
        ),
        (
            ('email','imap'),{
                'type':'textbox',
                'title':'Please enter your imap server information as "server:port"',
                'description':'In order to read your emails to you, I need to know the name and port of your imap server so I can log into your account',
                'validation':lambda IMAP : True if re.match('^.*:\d+$',IMAP) else False,
                'active':lambda : True if len(Values.email.address.strip())>0 else False
            }
        ),
        (
            ('email','password'),{
                'type':'password',
                'title':'Please enter the password to your email account',
                'description':'In order to read your emails to you, I will need the password to your imap server',
                'active':lambda : True if len(Values.email.address.strip())>0 and len(Values.email.imap.strip())>0 else False
            }
        )
    ]
)
```

So, in the first and second cases, it is using lambda functions to check if the value is a valid looking email address and a valid looking server:port string.

In the second and third cases, it is checking to make sure that the user provided values for the preceding settings. If they didn't, then skip the settings.

In a commandline session, this would look something like this:

```
Please enter your email address:// ?
This email address can be used by the check_email plugin to read your emails to you. It can also
be used by the notifications plugin to send notifications to you.

Please enter your email address:// aaron

I'm sorry, 'aaron' does not appear to be a valid email address. Just type a space and then enter
 to skip this field.

Please enter your email address: aaron// aaron@host.net

Please enter your imap server information as "server:port":// mail.host.net:993

Please enter the password to your email account:// **********
```

On the webserver, it would look like a regular web form:
```
<!DOCTYPE html>
<html lang="en-US">
<head>
<title>TestForm</title>
<meta charset="utf-8"/>
<script language="javascript">
function check(form,element){
  // I know this example doesn't really work as the onchange event
  // doesn't fire until after the current element loses focus,
  // meaning that any elements that get enabled by the "check"
  // function get skipped, but solving that problem is non-trivial
  // in javascript and probably justifies the use of a framework
  inputobjects=document.forms[form].getElementsByTagName("input");
  Enabled = true;
  alerttext="";
  for(i=0;i<inputobjects.length-1;++i){
    inputobjects[i].disabled=!Enabled;
    alerttext+=( inputobjects[i].name+" "+Enabled )+'\n';
    Enabled=inputobjects[i].value.length;
  }
  // alert(alerttext);
}
window.onload=function(){
  check('email');
};
</script>
</head>
<body>

<form name="email">
<label for="email_address" title="This email address can be used by the check_email plugin to read your emails to you. It can also be used by the notifications plugin to send notifications to you.">Please enter your email address</label><br />
<input id="email_address" name="address" type="text" onchange="check('email')"/><br />
<br />
<label for="email_imap" title="In order to read your emails to you, I will need to know the name and port of your imap server so I can log into your account">Please enter your imap server information as 'server:port'</label><br />
<input id="email_imap" name="imap" type="text" onchange="check('email')"/><br />
<br />
<label for="email_password" title="In order to read your emails to you, I will need the password to your imap server">Please enter the password to your email account</label><br />
<input id="email_password" type="password" name="password" onchange="check('email')"/><br />
<br />
<input type="submit" value="Submit"/>
</form>

</body>
</html>
```

I'd like to see the functions for checking validation and whether or not a function should be active implemented automatically as RESTful functions, so that the python functions can be used for the web interface rather than translating them to client side javascript.

Having this type of structure would also make it possible to check all the plugins as they are loaded and make sure all of their "settings" are represented in the profile (even if they contain only empty values). That way, the first time Naomi starts up after the user installs a new plugin, it can see that the user has not been asked about some values and present them to the user for configuration. Or if Naomi is running with a webserver interface, it might ask the user to visit the configuration page, where the new settings will be highlighted somehow.

Running Naomi with the --repopulate flag in command-line mode would still cause Naomi to loop through all core and plugin settings. Accordingly, we may want to implement an optional argument for repopulate that allows the user to specify a specific plugin (or "core") to run through the settings of, as we hope the list of plugins in use will eventually get very long (of course, many plugins will not require settings, so I don't know that this will ever become an issue).

I think Naomi should be allowed to go ahead and start even if the plugin itself can't be loaded because of the missing information, but a thread should ask for the information. Once the information has been updated, Naomi should attempt to re-initialize the plugin.

The goal is to move plugin settings out of Naomi and into the plugins themselves, allow for more flexibility in interfaces, and allow plugin developers to collect information they need for the plugin to function when it would be too time consuming to do that through the verbal interface.

## Context

I think this could be a big help to plugin developers, and attracting plugin developers is key to the success of this platform.

## Possible Implementation

See above

## Your Environment

* Version used: naomi-dev
* Environment name and version (e.g. PHP 5.4 on nginx 1.9.1): Python 3.5.3
* Server type and version: Raspberry Pi 3B+
* Operating System and version: Raspbian Stretch

**Comment by aaronchantrill** [comment]
I'm going to start by figuring out how to add the above structure and associated functions to the existing plugins and making sure I can extract the information from a populate.py program. Once I'm there, I will start working on a new, data driven version of populate.py. When that's done, I'll figure out how to check for unaddressed settings at each startup.

If anyone would like to start working on implementing the control types that I listed above in a friendlier, more curses-like way on the command line, I would be super grateful, as I'm not really interested in working on that and will just end up using the same hacky interface we are currently using. Also, @TuxSeb , this would be a great time to start figuring out how this could be implemented in HTML. I'm imagining a row of tabs, each tab representing a plugin, with tabs with unaddressed settings highlighted somehow, like a bold font or something. Clicking a tab would bring up a form listing all the settings associated with that plugin. I'm envisioning RESTful callbacks to the plugin developer's validate, active, default, and options python functions so that the form can be updated via AJAX as the user fills in details.

**Comment by G10DRAS** [comment]
The other good thing to implement is , instead of plain text yaml config file use a database like sqlite3 to store profile or configuration and provide a browser based interface to insert/update/delete configuration row.

something like this
https://github.com/G10DRAS/RoboCop/blob/master/images/OpenHAB.jpg

ofc there will be login credintials required to do operation on db.

**Comment by aaronchantrill** [comment]
Hey @G10DRAS thanks for responding. I don't see the advantage of switching the yaml hierarchical database structure to a relational database. The information just doesn't lend itself to tabular data, as far as I can see. There are, of course, other areas where it would make sense to use a relational database. I'm already using sqlite for storing training data, for example, but that is very tabular, with the same basic data stored for every request.

Can you elaborate on the advantages you see for using a relational db for storing configuration settings?

**Comment by G10DRAS** [comment]
Security is the main advantage. You can encrypt all your passwords, api keys and other sensitive information and store it in database.
Storing data in tabular format is not a big deal, you can store them with parent-child relationship.
Then it will be easy to Insert/update/delete data from database, either using a script or a web interface.

**Comment by aaronchantrill** [comment]
I still don't understand how adding a relational database engine simplifies encryption of passwords, api keys, etc. Python has a pretty robust set of Crypto modules and generally encryption/decryption is handled a the application layer, not within the database. I agree that someone needs to fix the storage of secrets ([Email username and password in plaintext #16](https://github.com/NaomiProject/Naomi/issues/16)) but don't understand how changing database engines helps that. It is also currently very easy to store and retrieve data using the PyYaml module. Can you detail how an RDB helps improve security in this case?

I would not call implementing a parent/child relationship within a relational database trivial. It would require writing an additional translation layer and introducing additional complexity, which would only be worthwhile if there is some real benefit. Now, it could simplify things to require that all configuration settings are stored at a single level and are not grouped so that it becomes just as easy to implement storage in a relational or hierarchical database, but that feels outside the scope of this specific issue, which is about detaching form descriptions from both the data and view layers.

Here are the benefits I see to sticking with the current implementation -- it is easy to read, edit, and share the profile or parts of the profile for debugging purposes, it only requires a single python module by way of dependencies, it automatically adapts for each user, and not changing it doesn't require any development time or rewriting documentation.

Thanks, Aaron

**Comment by aaronchantrill** [comment]
Resolved in Pull Request [#183 Replacement for Weather Plugin](https://github.com/NaomiProject/Naomi/pull/183). Still needs documentation, which I am working on.
